### PR TITLE
consensus: exception for special cases p2sh-assets.

### PR DESCRIPTION
**Pull request by fdoving**
Make exceptions for pre-p2sh-asset transactions that match previous checks.

### CODE REVIEW

**Review by HyperPeek — COMMENTED** [review]
First test show that this fixes the sync issue on mainnet. More testing needed before final approval. We need:
- true p2sh tx on mainnet to show it will be caught
- p2sh tx on testnet to show the test will let it through

**Comment by fdoving** [comment]
> First test show that this fixes the sync issue on mainnet. More testing needed before final approval. We need:
> 
> * true p2sh tx on mainnet to show it will be caught

I think trying to make this tx from the wallet with this version would be enough to test this.

> * p2sh tx on testnet to show the test will let it through
syncing testnet should be enough to proove this works, there  are already txes like this.

**Comment by HyperPeek** [comment]
Agree on 1. For 2 I am not worried, but technically any fix we made could have broken the p2sh ability post fork, so re-testing on testnet by sending through an updated wallet / node would proove it still works.

**Comment by HyperPeek** [comment]
Ok, just re-read -- you are right -- re-sync should already prove it and it works, so we should be fine.

**Comment by fdoving** [comment]
Synced from 0 successfully on ubuntu and freebsd.
Windows reindexed from 0 successfully.

Someone else should verify.

**Review by HyperPeek — APPROVED** [review]
Tested for some time and seems to work fine. Approve 9d3fde88334c8f49c68aae38a656912ee3643805 as we should have a broader base of testers which is tricky without this merged.

**Comment by ben-abraham** [comment]
The only potential pitfall I see is a slightly different tx being published between our testing now and pre-p2sh activation in the future.
IE: `OP_1 OP_0` instead, or any other currently-equally-valid combinations.
Those tx's would be considered valid now just the same as an `OP_0 OP_0`. I think we just remove the two OP_0 check's.

**Comment by HyperPeek** [comment]
I agree, that is a potential issue. So restrict as least as possible bf activation. This would leave OP_RVN_ASSET in [25] which we can enforce as it would fail subsequently anyway.

**Review by ben-abraham — APPROVED** [review]
Looks good

**Comment by HyperPeek** [comment]
This fix should work, as (*this)[25] == OP_RVN_ASSET still checks for p2pk Assets and will fail for p2sh (where OP_RVN_ASSET is in [23]). Also the magic "RVN" is in different location for p2sh and p2pk, so we could additionally check that for the p2sh position and reject that before activation.

**Comment by fdoving** [comment]
Need new solution. Need to set nScriptType to the correct value preP2SH.

**Pull request closed without merging**
